sdk/textanalytics/ai-text-analytics/src/transforms.ts

```diff
@@ -0,0 +1,346 @@
+ * is user-friendly.
+ *
+ * @param sentence - The sentence sentiment object to be converted.
+ * @param response - The entire response returned by the service.
+ * @returns The user-friendly sentence sentiment object.
+ * @internal
+ */
+function convertGeneratedSentenceSentiment(
+  { targets, assessments: _, ...rest }: GeneratedSentenceSentiment,
+  sentences: GeneratedSentenceSentiment[]
+): SentenceSentiment {
+  return {
+    ...rest,
+    opinions:
+      targets?.map(
+        // eslint-disable-next-line @typescript-eslint/no-shadow
+        ({ relations, ...rest }: SentenceTarget): Opinion => ({
+          target: rest,
+          assessments: relations
+            .filter((relation) => relation.relationType === "assessment")
+            .map((relation) => convertTargetRelationToAssessmentSentiment(relation, sentences)),
+        })
+      ) ?? [],
+  };
+}
+
+/**
+ * Converts a target relation object returned by the service to an assessment
+ * sentiment object where JSON pointers in the former are realized in the
+ * latter.
+ *
+ * @param targetRelation - The target relation object to be converted.
+ * @param response - The entire response returned by the service.
+ * @returns The user-friendly assessment sentiment object.
+ * @internal
+ */
+function convertTargetRelationToAssessmentSentiment(
+  targetRelation: TargetRelation,
+  sentences: GeneratedSentenceSentiment[]
+): AssessmentSentiment {
+  const assessmentPtr = targetRelation.ref;
+  const assessmentIndex: AssessmentIndex = parseAssessmentIndex(assessmentPtr);
+  const assessment: SentenceAssessment | undefined =
+    sentences?.[assessmentIndex.sentence].assessments?.[assessmentIndex.assessment];
+  if (assessment !== undefined) {
+    return assessment;
+  } else {
+    throw new Error(`Pointer "${assessmentPtr}" is not a valid Assessment pointer`);
+  }
+}
+
+/**
+ * @internal
+ */
+export function toEntityLinkingResult(
+  documents: TextDocumentInput[],
+  results: GeneratedEntityLinkingResult
+): EntityLinkingResult[] {
+  return transformDocumentResults(documents, results);
+}
+
+/**
+ * @internal
+ */
+export function toKeyPhraseExtractionResult(
+  documents: TextDocumentInput[],
+  results: GeneratedKeyPhraseExtractionResult
+): KeyPhraseExtractionResult[] {
+  return transformDocumentResults(documents, results);
+}
+
+/**
+ * @internal
+ */
+export function toEntityRecognitionResult(
+  documents: TextDocumentInput[],
+  results: GeneratedEntityRecognitionResult
+): EntityRecognitionResult[] {
+  return transformDocumentResults(documents, results);
+}
+
+/**
+ * @internal
+ */
+export function transformActionResult(
+  actionName: AnalyzeActionName,
+  input: TextDocumentInput[] | LanguageDetectionInput[],
+  response: AnalyzeResponse
+):
+  | EntityLinkingResult[]
+  | EntityRecognitionResult[]
+  | KeyPhraseExtractionResult[]
+  | SentimentAnalysisResult[]
+  | LanguageDetectionResult[] {
+  switch (response.kind) {
```

Comment:
Are these individual result types generated? If not, could they be specified such that they would automatically narrow by `"kind"`, removing the casting?

Reply:
The result types themselves are not generated but the way the generated response is typed makes it hard to do anything useful with the narrowed type. In particular, without casting, you will get this error:
> Property 'results' does not exist on type 'AnalyzeTextTaskResult | EntityLinkingTaskResult'.
  Property 'results' does not exist on type 'AnalyzeTextTaskResult'.ts(2339)

`AnalyzeTextTaskResult` has the `kind` field only and is part of the union. Narrowing will include it in the result type because its `kind` field captures all options. 

P.S. There is a discussion around alternatives to model swagger inheritance in https://github.com/Azure/autorest.typescript/issues/1132.

Reply:
CADL also solves this since you can use both `extends` and spread syntax when composing models and they have a union type (including [named unions](https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#named-unions))